procasl/preprocessing.py

```diff
@@ -0,0 +1,488 @@
+# TODO: keep only classes, move functions to _utils, rename to preprocessor
+import os
+
+import numpy as np
+from scipy.io import savemat
+import nibabel
+import nipype.interfaces.spm as spm
+from nipype.interfaces import fsl
+fsl.FSLCommand.set_default_output_type('NIFTI')
+from nipype.interfaces.base import BaseInterface, \
+    BaseInterfaceInputSpec, traits, File, TraitedSpec, Directory
+from nipype.utils.filemanip import split_filename
+from nipype.interfaces.base import InputMultiPath, OutputMultiPath
+
+from procasl.spm_internals import params_to_affine, spm_affine
+from procasl._utils import check_images
+
+
+def add_prefix(prefix, in_file):
+    """Adds a prefix to a filename
+
+    Parameters
+    ----------
+    prefix : str
+        Prefix to append to the filename
+
+    in_file : str
+        Input file name.
+
+    Returns
+    -------
+    out_file : str
+        Output file name
+    """
+    out_file = os.path.join(os.path.dirname(in_file),
+                            prefix + os.path.basename(in_file))
+    return out_file
+
+
+def get_scans_number(in_file):
+    """Return the number of scans for a 4D image."""
+    image = nibabel.load(in_file)
+    data = image.get_data()
+    if data.ndim != 4:
+        raise ValueError("Expect a 4D image, got a {}D".format(data.ndim))
+
+    return data.shape[-1]
+
+
+def select_scans(in_file, selected_scans, convert_3d=True, out_file=None):
```

Comment:
keep consistent behavior with `_utils.remove_nan`: decide once for all whether give output filename or impose it